src/Native/Runtime/amd64/AllocFast.S

```diff
@@ -0,0 +1,260 @@
+
+        // If the object is bigger than RH_LARGE_OBJECT_SIZE, we must publish it to the BGC
+        mov         esi, [rbx + OFFSETOF__EEType__m_uBaseSize]
+.att_syntax
+        cmp         $RH_LARGE_OBJECT_SIZE, %rsi
+.intel_syntax noprefix
+        jb          New_SkipPublish
+        mov         rdi, rax            // rdi: object
+                                        // rsi: already contains object size
+        call        C_FUNC(RhpPublishObject) // rax: this function returns the object that was passed-in
+New_SkipPublish:
+
+        .cfi_remember_state
+        POP_COOP_PINVOKE_FRAME
+        ret
+
+        .cfi_restore_state
+NewOutOfMemory:
+        // This is the OOM failure path. We're going to tail-call to a managed helper that will throw
+        // an out of memory exception that the caller of this allocator understands.
+
+        mov         rcx, rsi            // EEType pointer
+        xor         edx, edx            // Indicate that we should throw OOM.
+
+        POP_COOP_PINVOKE_FRAME
+
+        jmp         C_FUNC(RhExceptionHandling_FailedAllocation)
+NESTED_END RhpNewObject, _TEXT
+
+
+// Allocate one dimensional, zero based array (SZARRAY).
+//  RDI == EEType
+//  ESI == element count
+NESTED_ENTRY RhpNewArray, _TEXT, NoHandler
+        // we want to limit the element count to the non-negative 32-bit int range
+        cmp         rsi, 07fffffffh
+        ja          ArraySizeOverflow
+
+        push_nonvol_reg rbx
+        push_nonvol_reg r12
+        push_register rcx       // padding
+
+        mov         rbx, rdi    // save EEType
+        mov         r12, rsi    // save element count
+
+        // rax = GetThread()
+        INLINE_GETTHREAD
+
+        mov         rcx, rax    // rcx = Thread*
+
+        // Compute overall allocation size (align(base size + (element size * elements), 8)).
+        movzx       eax, word ptr [rbx + OFFSETOF__EEType__m_usComponentSize]
+        mul         r12
+        mov         edx, [rdi + OFFSETOF__EEType__m_uBaseSize]
+        add         rax, rdx
+        add         rax, 7
+        and         rax, -8
+
+        // rax == array size
+        // rbx == EEType
+        // rcx == Thread*
+        // r12 == element count
+
+        mov         rdx, rax
+        add         rax, [rcx + OFFSETOF__Thread__m_alloc_context__alloc_ptr]
+        jc          RhpNewArray_RarePath
+
+        // rax == new alloc ptr
+        // rbx == EEType
+        // rcx == Thread*
+        // rdx == array size
+        // r12 == element count
+        cmp         rax, [rcx + OFFSETOF__Thread__m_alloc_context__alloc_limit]
+        ja          RhpNewArray_RarePath
+
+        mov         [rcx + OFFSETOF__Thread__m_alloc_context__alloc_ptr], rax
+
+        // calc the new object pointer
+        sub         rax, rdx
+
+        mov         [rax + OFFSETOF__Object__m_pEEType], rbx
+        mov         [rax + OFFSETOF__Array__m_Length], r12d
+
+        .cfi_remember_state
```

Comment:
Yes, cfi_remember_state does not work as expected. I have tweaked the directives to make them work (verified by checking stacktrace for each instruction under debugger).